### PR TITLE
Joining an existing Link session should adopt the session tempo, not force the device tempo

### PR DESCRIPTION
**Pull request by Dewb**
### background/motivation

The [Link test plan](https://github.com/Ableton/link/blob/master/TEST-PLAN.md) has two clauses about the expectations for how joining a session should affect the app/device tempo:

[TEMPO-2](https://github.com/Ableton/link/blob/master/TEST-PLAN.md#tempo-2-opening-an-app-with-link-enabled-should-not-change-the-tempo-of-an-existing-link-session): Opening an app with Link enabled should not change the tempo of an existing Link session.
[TEMPO-5](https://github.com/Ableton/link/blob/master/TEST-PLAN.md#tempo-5-enabling-link-does-not-change-apps-tempo-if-there-is-no-link-session-to-join): Enabling Link does not change app's tempo if there is no Link session to join.

So the behavior needs to be different based on whether the session has other participants or not. Luckily, this behavior is already implemented in the Link library; in order to take advantage of it, we just need to ~~construct a new Link object when enabling Link~~ manipulate the enable state, and make sure tempo requests are processed before we change the enable state.

### primary change: enable/disable Link when it is selected/deselected as a source

The existing behavior is to create and join a Link session at norns startup, and then switch it to be the clock source (and set the tempo) when Link is enabled from the params menu. The primary change in this PR is to only enable the Link client when Link becomes the clock source, and to send the tempo change request before Link becomes the clock source.

This has a nice consequence in that other Link clients that report number of connected devices will now only count norns when it's using Link for sync, rather than any time it's on the network regardless of clock source.

~~This also has a not-nice consequence: the Link "start/stop sync" feature, which norns previously always enabled, now exhibits behavior that doesn't seem correct or desirable. With start-stop sync always enabled, it's easy to get to a state where you have no way of re-enabling transport from norns, you have to kick the session from another client (or run `_norns.clock.start()` from maiden.)~~

### ~~secondary change: allow user to toggle Link start/stop sync~~

~~To mitigate this, this PR also exposes the Link 3.0 start/stop sync enable/disable state in the menu, and defaults it to disabled. This way Link behaves more like the other clock sources, and it's harder to inadvertently stop the transport by switching sources.~~

~~Start-stop sync is currently tricky to implement correctly because the system does not currently have a clock-source-independent, script-independent notion of whether the global transport is running or not. Having such a state isn't an explicit requirement for Link clients, but reading between the lines of the test plan, that assumption does seem to be in there. A follow-up PR could add a global transport state to cover the edge cases when switching Link on and off with start/stop sync enabled.~~

### CODE REVIEW

**Comment by Dewb** [comment]
I think this covers 1-4 in #1376. Item 5 would be a good fit with that hypothetical followup that adds a global transport state.

**Comment by Dewb** [comment]
it's also possible that I'm missing something and start/stop sync is totally possible to implement ergonomically without global transport state!

**Review by artfwo — CHANGES_REQUESTED** [review]
it should not be necessary to expose join_session/leave_session in the API, as well as initializing clock in clock_link_run to solve these issues:

> 
> TEMPO-2: Opening an app with Link enabled should not change the tempo of an existing Link session.
> TEMPO-5: Enabling Link does not change app's tempo if there is no Link session to join.

link already does this internally, but it doesn't quite work in norns because of the [global tempo follower](https://github.com/monome/norns/blob/main/lua/core/clock.lua#L281) that basically synchronizes internal tempo with midi/link/crow.

a less complicated solution would be removing the tempo follower altogether (as it's broken anyway) and replacing it with fallback tempo values for midi and crow, so setting the clock tempo parameter will only change fallback values for those sources. with this, we won't need to track the tempo and will be able to piggyback on link's default behaviors without additional API changes (also see https://github.com/monome/norns/issues/1374 for discussion and some context).

happy to review the start/stop sync change separately, could you please bring it up as a different PR?

**Comment by Dewb** [comment]
> it should not be necessary to expose join_session/leave_session in the API, as well as initializing clock in clock_link_run to solve these issues:

Ah yes, I see the join behavior is actually triggerable through Controller::enable, and there was no need to expose join/leave to the lua layer. I can definitely simplify this quite a bit.

Thanks, I hadn't seen #1374 and I wasn't aware the tempo follower behavior was implicated in this issue. I don't quite follow how it's preventing us from letting Link decide how to handle the join tempo though? It looks like the fundamental issue is that changing the clock source in params calls `clock.link.set_tempo()`, which gets passed to the Link client as if it were a user-initiated change, which it is not.

**Comment by artfwo** [comment]
> Thanks, I hadn't seen #1374 and I wasn't aware the tempo follower behavior was implicated in this issue. I don't quite follow how it's preventing us from letting Link decide how to handle the join tempo though? It looks like the fundamental issue is that changing the clock source in params calls `clock.link.set_tempo()`, which eventually gets passed to the Link client as if it were a user-initiated change, which it is not.

you're right, in this case it doesn't (sorry, I haven't had coffee yet :)) but I think this code should be removed too.

generally, we should let link handle all the joining internally, as it implements the expectations correctly by default and maybe decouple the clock sources a bit to let that happen.

**Comment by Dewb** [comment]
Drastically simplified the changes. Yes, most of that stuff was unnecessary! Manipulating the enable state and making sure we set the local tempo before enabling Link is enough to satisfy TEMPO-2 and TEMPO-5, and have the correct number of peers show up in other clients.

I'm happy to keep iterating on this, or if you ultimately don't think this direction is promising, I can trim it down to just the start/stop state sync exposure.

**Comment by artfwo** [comment]
> I'm happy to keep iterating on this, or if you ultimately don't think this direction is promising, I can trim it down to just the start/stop state sync exposure.

Yes, please go ahead. I'm going to have a closer look at the changes later this week, but a quick question for now - is clock start/stop sync preserved correctly after norns is reset?

**Comment by Dewb** [comment]
> is clock start/stop sync preserved correctly after norns is reset?

A sleep/reboot cycle preserves the state, yes. `SYSTEM > RESET` resets it to default (disabled.)

**Comment by Dewb** [comment]
Split this into 2 PRs, this one for the join behavior, the other for adding the parameter for start/stop sync.

File: matron/src/clocks/clock_link.c [hunk 1]
```diff
@@ -63,6 +63,8 @@ static void *clock_link_run(void *p) {
                 clock_link_shared_data.requested_tempo = 0;
             }
 
+            ableton_link_enable(link, clock_link_shared_data.enabled);
```
Comment (artfwo) [inline matron/src/clocks/clock_link.c:66]: this is a [thread-safe method](https://github.com/Ableton/link/blob/master/include/ableton/Link.hpp#L98), so technically we don't need to pass the flag through clock_link_shared_data and just call ableton_link_enable with the appropriate parameter from clock_set_source.

Comment (Dewb) [reply matron/src/clocks/clock_link.c:66]: The nice thing about having the `ableton_link_enable()` call inside the mutex-guarded link thread is that we can guarantee that any pending tempo change requests are processed before we call enable.

Comment (artfwo) [reply matron/src/clocks/clock_link.c:66]: how does mutex ensures that? the only thread where these requests could come from is the single matron thread, so there can't be any race conditions around tempo.

Comment (Dewb) [reply matron/src/clocks/clock_link.c:66]: this is totally theoretical, just thinking out loud, but:
* change clock source from midi to link, param handler for clock_source runs
* handler calls `clock.link.set_tempo()`. link thread is sleeping so this grabs mutex, updates shared data structure, and releases mutex.
* link thread wakes up and starts running
* handler next calls `clock.set_source` and directly sets enable state
* link thread reaches the block `if (clock_link_shared_data.requested_tempo > 0)` and processes tempo change. since link has already been enabled, this is treated as a user-initiated tempo change and if there are other clients in the session, this violates the TEMPO-2 testcase.

Comment (Dewb) [reply matron/src/clocks/clock_link.c:66]: in other words, we need the disabled -> enabled transition to happen here on line 66, after the tempo change request has been committed.

**Review by artfwo — APPROVED** [review]
thanks, i think the approach described in https://github.com/monome/norns/issues/1374#issuecomment-854067186 could be implemented as a follow-up, so let's merge this first.
